File: pkg/R/cholesky.R
## ID: cholesky.R, last updated 2020-09-23, F.Osorio

chol2inverse <- function(x)
{ ## compute weighted mean and covariance using an online algorithm
  if (is.data.frame(x))
    x <- as.matrix(x)
  if (!is.matrix(x))
    stop("supply a matrix-like 'x'")
  if (!is.numeric(x))
    stop("argument x is not a numeric matrix")
  if (!is.upper.tri(x))
    stop("argument x is not an upper triangular matrix")

  storage.mode(x) <- "double"
  ldx <- p <- nrow(x)
  job <- 1 # 'x' is upper triangular

  y <- .C("chol_inverse",
          x = x,
          ldx = as.integer(ldx),
          p = as.integer(p),
          job = as.integer(job),
          info = as.integer(0))

  if (y$info > 0)
    stop("inverse of 'x' could not be computed")

  y <- matrix(y$x, nrow = p, ncol = p)
  y
}

File: pkg/src/cholesky.c
/* ID: cholesky.c, last updated 2022-02-10, F.Osorio */

#include "fastmatrix.h"

/* Cholesky decompositions */

void
chol_dcmp(double *a, int *lda, int *p, int *job, int *info)
{ /* wrapper to 'FM_chol_decomp' */
  FM_chol_decomp(a, *lda, *p, *job, info);
}

void
chol_update(double *r, int *ldr, int *p, double *x)
{ /* update the Cholesky decomposition */
  int n = *p;
  double *c, *s;

  /* cosines and sines of transforming rotations */
  c = (double *) Calloc(n, double);
  s = (double *) Calloc(n, double);

  /* update upper triangular matrix */
  for (int j = 0; j < n; j++) {
    double xj = x[j];

    /* apply the previous rotations */
    for (int i = 0; i < j; i++) {
      double aux = *(r + i + j * n) * c[i] + xj * s[i];
      xj = xj * c[i] - *(r + i + j * n) * s[i];
      *(r + i + j * n) = aux;
    }

    /* compute next rotation */
    BLAS1_rotg((r + j + j * n), &xj, (c + j), (s + j));
  }

  Free(c); Free(s);
}
